### PR TITLE
#14 Add configurable derived Encoder and Decoder

### PR DESCRIPTION
**Pull request by jatcwang**
- Support the same configuration as circe-generic-extras (Copied the Configuration class from circe-generic-extras project as well as JsonKey)

- [x] Support `JsonKey` annotation
- [x] Add equivilence test with circe-generic-extras (also add extras tests for extra safety)
- [x] Various cleanup
- [x] Update README

We should also add some benchmark in a future PR against circe-generic-extras to ensure the performance is at least on par.

### CODE REVIEW

**Comment by jatcwang** [comment]
I see you want to mirror circe-generic and circe-generic-extras. I'm not sure how much gain there is in doing so. Here are the benefits of keeping them in the same module:

* Share the same implementation
* JsonKey support for non-configured encoder/decoders

WDYT

**Comment by vpavkin** [comment]
@jatcwang 
Sure, we can start with a single module. Ping me when you feel ready to merge :)

**Comment by jatcwang** [comment]
We should run some benchmarks against circe-generic and circe-generic-extras before release to make sure there aren't a major performance regression, so that this can be a drop-in replacement. Splitting implementation (i.e. different modules) may be one solution to improve performance, if needed.

WDYT about code organization / usage right now?

```
import io.circe.magnolia.configured.encoder.semiauto.deriveConfiguredMagnoliaDecoder

object MyClass {
  implicit val config = ...
  val decoder: Decoder[MyClass] = deriveConfiguredMagnoliaDecoder[MyClass]
}
```

**Comment by jatcwang** [comment]
This PR is ready for some feedback.

Currently generic-extras' semiauto `deriveEncoder` derives an `ObjectEncoder` instead of `Encoder`, but there's an edge case where if one of the subtype has a custom Encoder instance that does not encode to a JsonObject, the Encoder -> Decode round trip law is broken as demonstrated by the following code snippet. 
```
package io.circe.magnolia.configured

import org.scalatest.FreeSpec

import io.circe.generic.extras.semiauto._
import io.circe.generic.extras.{Configuration => GeConfiguration}
import io.circe._

class BlahSpec extends FreeSpec {
  "asdf" in {
   // Interestingly, this encodes as if discriminator is not configured, I guess to enforce a JsonObject result 
    val sj = Sealed.encode(Too(List(1, 2, 3)))
    println(Sealed.encode(Zoo(2)))
    println(sj)
    val de = Sealed.decode(sj.hcursor)
    println(de)
  }
}

sealed trait Sealed

case class Zoo(i: Int) extends Sealed

case class Too(ss: List[Int]) extends Sealed

object Too {
  // One of the subtype is not encoded to a case class
  implicit val encode: Encoder[Too] = Encoder[List[Int]].contramap(_.ss)
  implicit val decode: Decoder[Too] = Decoder[List[Int]].map(Too(_))
}

object Sealed {
  implicit val c = GeConfiguration.default.withDiscriminator("type")
  implicit val encode = deriveEncoder[Sealed]
  implicit val decode = deriveDecoder[Sealed]
}
```

I want to have magnolia's derivation derive `ObjectEncoder` instead of `Encoder` as having the former has proven to be very useful in my experience. However, if the user wants to use a custom encoder for one of the subtypes, it will need to be an `ObjectEncoder` instance instead of `Encoder`, as magnolia is locked to searching for one Typeclass (it's not as flexible as shapeless)

I think this is what we should do, for two reasons:

* Current generic-extras behaviour is broken (as demonstrated above). 
* Derivation should cater for the most common use cases, if you want to do special edge-casey things like encode one of your subtypes as a JSON array, you should handwrite your own encoders and decoders.

However, **This is a deviation in behaviour from our reference libraries**. WDYT

**Comment by vpavkin** [comment]
Hi, @jatcwang.
Thanks, I'll review it over the next several days (I'm at f.by conference during the weekend).

As to the problem with `ObjectEncoder` -- I have thought on this one as well, and even submitted an issue to magnolia to think about a solution: https://github.com/propensive/magnolia/issues/107

IDK if proposed feature will eventually land, but it solves the problem described in a rather generic way.

**Comment by jatcwang** [comment]
I think Jon will be pretty busy with Fury for a while. I'm wondering whether we should just require ObjectEncoder throughout for now, with documented differing behaviour from circe-generic. This is a tighter restriction which we relax when/if the magnolia change you propose is implemented.

(IMO, I think circe-generic should require ObjectEncoder from subtypes as well, otherwise the encode -> decode roundtrip law is broken)

**Comment by vpavkin** [comment]
I think I get your point, but I'll need more time and a more peaceful environment to think about it all :)
Hope to come up with something after the weekend.

**Comment by vpavkin** [comment]
btw, I found the snippet above quite enlightening and scary, thanks for it:)

**Comment by jatcwang** [comment]
Oops I totally forgot about `combine` and primitive types 😅 So no it won't work unfortunately.
It's possible to do some hack involving implicit conversions so we can derive `ObjectEncoder`s but I don't think the performance and code complexity cost is worth it.

Let's stick with `Encoder` for now until magnolia allows for different inner and outer typeclass types.

I'll do some further cleanup and update the docs. Will let you know when it's ready :) Thanks

File: README.md [hunk 1]
```diff
@@ -57,6 +57,27 @@ val decoder = deriveMagnoliaDecoder[Foo]
...
+case class User(firstName: Int, lastName: String)
+
+implicit val configuration: Configuration = Configuration.default.withSnakeCaseMemberNames
+
+val encoder = deriveConfiguredMagnoliaEncoder[User]
+val decoder = deriveConfiguredMagnoliaDecoder[User]
+```
+
+To avoid constantly needing to provide/import an implicit `Configuration` to derive Encoder/Decoders, you can hardcode the configuration by defining your own version of deriver. 
+See `HardcodedDerivationSpec` for an example of how to do this. (In fact, the default auto and semiauto derivers are hardcoded to use the default Configuration)
```
[the remark this answers is not in the document]
Comment (jatcwang) [reply README.md]: done :)

**Comment by vpavkin** [comment]
:rocket:

**Comment by vpavkin** [comment]
released in 0.4.0

**Pull request merged**